flow/project.py

```diff
@@ -2317,6 +2560,59 @@ def _guess_cmd(func, name, **kwargs):
                     cmd=_guess_cmd(func, name, **params), **params)
                 self._operation_functions[name] = func
 
+    @classmethod
+    def make_group(cls, name, run_cmd=None, exec_cmd=None, directives=None, options=None):
+        # Gets the relative filepath of the function caller
+        filepath = inspect.stack()[1][1]
+        cls._GROUPS.append({'name': name, 'run_cmd': run_cmd,
+                            'exec_cmd': exec_cmd, 'directives': directives,
+                            'options': options, 'group_path': filepath})
+
+        # Create decorator that adds a _flow_group label to operations.
+        def add_to_group(func):
+            if hasattr(func, '_flow_group'):
+                func._flow_group.append(add_to_group.group_name)
+            else:
+                func._flow_group = [add_to_group.group_name]
+            return func
+        # Specify group name
+        add_to_group.group_name = name
+        return add_to_group
+
+    def _register_groups(self):
+        "Register all groups and add the correct operations to each."
+        groups = []
+        # Gather all groups from class and parent classes.
+        for cls in type(self).__mro__:
+            groups.extend(getattr(cls, '_GROUPS', []))
+
+        # Initialize all groups without operations
+        for group in groups:
+            # Raise error for duplicate groups
+            if group['name'] in self._groups:
+                raise ValueError(
+                    "Repeat definition of group with name " +
+                    "'{}'.".format(group['name']))
+            # Create group
+            self._groups[group['name']] = FlowGroup(**group)
+
+        # Add operations to group
+        # Currently recreating operation list. There may be a better way to do
+        # this.
+        operations = []
+        for cls in type(self).__mro__:
+            operations.extend(getattr(cls, '_OPERATION_FUNCTIONS', []))
+        for (name, func) in operations:
+            if hasattr(func, '_flow_group'):
+                for group in func._flow_group:
+                    self._groups[group].add_operation(name,
+                                                      self._operations[name])
+
+    def _get_operation_path(self, op_name):
```

Comment:
It seems to me that wherever we call this function, we actually have a reference to the actual operation. Why are we looking up by name here?

Reply:
We need the path which is only available by looking at the function. Since `FlowOperations` do not hold references to their functions, the only way to obtain the path is in iterating through the names.

It turns out that our talk about using the `FlowProject's` knowledge of its location would only work if we added logic to it's `__init__()` function. Similar to the logic found in the `make_groups` class method.

I would be fine with adding the logic to `FlowProject` but it seemed out of the scope of this pull request.

Reply:
Ok, I just wanted to make sure that there was good reason for this.